easybuild/easyconfigs/r/RFdiffusion/RFdiffusion-1.1.0-foss-2022a-CUDA-11.7.0.eb

```diff
@@ -1,61 +1,189 @@
+# _models_url + / + X + / + F
+# X is not a checksum. According checksums are found in _model_checksums.
+_models = [
+    ("Base_ckpt.pt", "6f5902ac237024bdd0c176cb93063dc4"),
+    ("Complex_base_ckpt.pt", "e29311f6f1bf1af907f9ef9f44b8328b"),
+    ("Complex_Fold_base_ckpt.pt", "60f09a193fb5e5ccdc4980417708dbab"),
+    ("InpaintSeq_ckpt.pt", "74f51cfb8b440f50d70878e05361d8f0"),
+    ("InpaintSeq_Fold_ckpt.pt", "76d00716416567174cdb7ca96e208296"),
+    ("ActiveSite_ckpt.pt", "5532d2e1f3a4738decd58b19d633b3c3"),
+    ("Base_epoch8_ckpt.pt", "12fc204edeae5b57713c5ad7dcb97d39"),
+    ("Complex_beta_ckpt.pt", "f572d396fae9206628714fb2ce00f72e"),
+    ("RF_structure_prediction_weights.pt", "1befcb9b28e2f778f53d47f18b7597fa"),
+]
+_models_url = 'http://files.ipd.uw.edu/pub/RFdiffusion'
+_schedules_url = 'https://files.ipd.uw.edu/krypton/'
+_schedules_fn = 'schedules.zip'
+
+_model_checksums = [
+    # Base_ckpt.pt
+    '0fcf7d7c32b4848030aca3a051e6768de194616f96ba6c38186351a33bfc6eca',
+    # Complex_base_ckpt.pt
+    '76e4e260aefee3b582bd76b77ab95d2592e64f00c51bf344968ab9239f3250bc',
+    # Complex_Fold_base_ckpt.pt
+    '0ac3b4024aea811078cec41482528291d6d7d7084bf8190ec118f54642fb81a1',
+    # InpaintSeq_ckpt.pt
+    '3b71b2b954e87d46b75a88ba64e0420fbf27f592604b10b6c3561b8c8ab70ab6',
+    # InpaintSeq_Fold_ckpt.pt
+    '51849c9fe64c16a38fe41c75db76abe044e4d3493926f6cfd29a5bde0331b7cc',
+    # ActiveSite_ckpt.pt
+    'beca1f672049161df0bc6a2d2523828f19fd9c8a2b449988e246dde42e7ea986',
+    # Base_epoch8_ckpt.pt
+    'b8e5d57f0b8a8f8cb30779c106b75210b46a914a4d19fb180676ae647f5ae23d',
+    # Complex_beta_ckpt.pt
+    '5a0b1cafc23c60b1aabcec1e49391986ac4fd02cc1b6b4cc41714ca9fe882e9e',
+    # RF_structure_prediction_weights.pt
+    '6414e9e60b0b01011e5a182def40b4e6de4e137554c887b2916d43566733ed95',
+]
+_schedules_checksum = '9fbb47ddb92f1c0cf234bec39ba6ba64d179501fbff53a4d10b4e55afd75e44a'
+
+_pysp = '%(installdir)s/lib/python%(pyshortver)s/site-packages'
+_models_dst = '%s/models/' % _data_dir
+_schedules_dst = '%s/schedules/' % _data_dir
+_get_models = '&&'.join(["cd %s && wget -nc %s/%s/%s" % (_models_dst, _models_url, _lv, _ln) for _ln, _lv in _models])
+# wget: do not overwrite
+_get_models += '&&'
+_get_models += '&&'.join(["[ $(shasum -a 256 %s/%s |awk '{print $1}') == %s ]||{echo wrong checksum for %s;exit 1;}" %
+                         (_models_dst, _ln, _lcs, _ln) for (_ln, _), _lcs in zip(_models, _model_checksums)])
+_get_schedules = 'cd %s && wget %s/%s ' % (_schedules_dst, _schedules_url, _schedules_fn)
+_get_schedules += "&& [ $(shasum -a 256 %s |awk '{print $1}') == %s ]" % (_schedules_fn, _schedules_checksum)
+_get_schedules += '&& unzip -n %s' % _schedules_fn  # unzip: do not overwrite
+_get_schedules += '&& mv RFdiffusion/schedules/* . && rm RFdiffusion %s -rf' % _schedules_fn
+
+
+# uncomment for installing models and schedules as components.
+# Then comment out _get_schedules and _get_models in postinstallcmds
```

Comment:
what's the difference between the components and _get_schedules/_get_models?
i would prefer to use the best method and remove the other one to keep things clean

Reply:
@smoors Using components would archive the models and would allow to reinstall RFdiffusion even if the models were not available for download any more. I'd prefer this option since RFdiffusion is not usable without the models.
However, I understand @lexming's concerns about to store the models data (3.9GB) at least twice, in the module directory and the sources directory.
Therefore I kept the components approach commented out.

Reply:
Which solution do you prefer and which should I delete?

Reply:
hm, i agree with you that having to re-download on reinstall is suboptimal, but i also agree with  @lexming that storing the models twice is suboptimal.

a possible solution is to create a custom easyblock that checks the installdir and does not delete/re-download the models if they are already present and if their checksums are correct.
to make this configurable, we could define an extra parameter in the easyblock, e.g. `keep_models_if_installed`.
this is similar to easyblock `Tarball`, which allows merging an installation into an existing installation instead of wiping it first.

what do you think?

Reply:
For long term it is probably better to place the data into a separate module RFdiffusion_data, which can be reused for minor updates of RFdiffusion, if the data does not change.
It might be good to have a generic easyblock for such data modules, because we can expect more of them with all the upcoming AI packages. (see also #19157 Model-Angelo_data, #19240 relion-classranker_data + relion-blush_data).
There was also a brief discussion at https://github.com/easybuilders/easybuild/wiki/Conference-call-notes-20230719

Reply:
that's indeed a better solution long-term.
i'll try to give a stab at a generic data easyblock, but feel free to do this yourself if you're up for it.

in the meantime, i suggest to already go ahead with creating a separate module RFdiffusion_data. you should be able to use the `Bundle` easyblock (which does not install anything), and download the models in a `postinstallcmds`, as you have done here.

Reply:
I'll have a look on `RFdiffusion_data` in the next few days. `RFdiffusion-1.1.0_data_paths.patch` requires an update allowing to read the models and schedules from elsewhere defined in a environment variable.